Remove gunicorn daemonize for api-server

jason810496 opened the pull request:
related: #52581

## Why

When I attempt to remove `gunicorn` dependency in #52581, I found that we are still using `gunicorn` for daemonization.

## What

- Replace `gunicorn.daemonize` with `run_command_with_daemon_option`, this also align the behavior of `--daemon` option for scheduler, triggerer ... and **api-server**
- Add test for api-server with `--daemon` as well
  - Rename `_CommonCLIGunicornTestClass` to `_CommonCLIUvicornTestClass`, since we are using `uvicorn` instead of `gunicorn` for api-server now.
- Replace `[api/access_logfile]` config with `[api/log_config]`, detail in  https://github.com/apache/airflow/pull/52860#discussion_r2185721521

jason810496 committed:
Remove gunicorn daemonize for api-server

jason810496 committed:
Rename _CommonCLIUvicornTestClass

jason810496 committed:
Add test for api-server with daemon option

jason810496 commented on airflow-core/src/airflow/cli/commands/api_server_command.py:
```diff
@@ -40,6 +40,40 @@
+        timeout_graceful_shutdown=worker_timeout,
+        ssl_keyfile=ssl_key,
+        ssl_certfile=ssl_cert,
+        access_log=access_logfile,  # type: ignore[arg-type]
```
The `access_log` argument type of `uvicorn.run` is annotated as  `access_log: bool = True,`, but we are accepting `access_logfile` as `str` and pass to `uvicorn.run`.

https://github.com/encode/uvicorn/blob/master/uvicorn/main.py#L492

https://github.com/apache/airflow/blob/7a44f9ada56175db83b69902fc6195bcbd419084/airflow-core/src/airflow/config_templates/config.yml#L1374-L1380

Do we need to cleanup the `access_logfile` config?
Since no matter what user set, `access_logfile` will be cast to `True` and pass to `uvicorn.run`.

potiuk replied:
Hm. But thos was supposed to be "string" -> path to the file  ? 

Can we bring it back to have the same semantics? I.e. if "file_path" - log to that path, but if "-" log to stdout?

potiuk replied:
I think we can pass dict logconfig with logging configuration to write to the log file ?

potiuk replied:
(ah - for daemonize "-" / stdout does not really work, but passing a path to file where to write logs makes perfect sense I think?

jason810496 replied:
After tracing the history of Airflow 2 and gunicorn:
- **gunicorn**
  - `--access-logfile`: The Access log file to write to and `-` means `stdout`
  - https://github.com/benoitc/gunicorn/blob/a86ea1e4e6c271d1cd1823c7e14490123f9238fe/docs/source/settings.rst#accesslog
- **uvicorn**
  - `--log-config`: Logging configuration file that is valid for `logging.config.fileConfig`
  - https://www.uvicorn.org/settings/#production

The `--access-logfile` option for `api_server_command` is inherited from Airflow 2 ( we were still using `gunicorn`  then ).

To sum up, maybe we can deprecate the `access-logfile` config  ( since it's legacy stuff of `gunicorn`, we are not using anymore ) and add the `log-config` (  for `uvicorn` now ) and default as `None`?  This remain current behavior of api-server and provide the flexibility for user to set to logging related config for `uvicorn`.

potiuk replied:
Yeah. lets.

jason810496 reviewed (commented):
`access_log=access_logfile,  # type: ignore[arg-type]` this line is a strange part for me.
> I add `# type: ignore[arg-type]` to ingore the mypy warning for now.

potiuk commented on airflow-core/src/airflow/cli/commands/api_server_command.py:
```diff
@@ -40,6 +40,40 @@
+):
+    """Run the API server."""
+    log.info(
+        textwrap.dedent(
```
Nice!

potiuk commented on airflow-core/src/airflow/cli/commands/api_server_command.py:
```diff
@@ -40,6 +40,40 @@
+    # get ssl cert and key filepaths here instead of passing them as arguments to reduce the number of arguments
+    ssl_cert, ssl_key = _get_ssl_cert_and_key_filepaths(args)
+
+    setproctitle(f"airflow api_server -- host:{args.host} port:{args.port}")
```
I remember there was a problem with setproctitle on MacOS because of forking/threading clash and I think we have conditional code to not use setproctitle on `darwin`.

jason810496 replied:
Yes, we also handle this special case for task_runner  in task-sdk  as well. https://github.com/apache/airflow/pull/45124/files Nice catch!

jason810496 committed:
Fix setproctitle on MacOS

jason810496 committed:
Replace api/access_logfile with api/log_config

jason810496 committed:
Fix [api/log_config] cli_config

jason810496 committed:
Remove logging and unsed option in api_server_command

- the logging before daemonization is not logging the correct PID so
  remove it
- should_setup_logging is default as False

jason810496 committed:
Fix test_run_command_daemon

jason810496 reviewed (commented):
Thanks @potiuk and @pierrejeambrun  for the review!
Just resolved all comments and fixed the CI as well.

potiuk commented on airflow-core/src/airflow/config_templates/config.yml:
```diff
@@ -1371,13 +1371,14 @@ api:
       example: ~
       default: "120"
-    access_logfile:
+    log_config:
```
I think that one needs significant newsfragment. It's really solving a deployment issue - so not a "breaking" change - it clarifies the options that **SHOULD** be set, but it's disruptive enough that should be mentioned as significant change in the changelog.

jason810496 replied:
Sound reasonable, just add significant newsfragment rst file.

potiuk reviewed (approved):
Nice. One NIT about updating newsfragment.

jason810496 committed:
Add significant rst

jedcunningham commented on airflow-core/src/airflow/configuration.py:
```diff
@@ -373,6 +373,7 @@ def sensitive_config_values(self) -> set[tuple[str, str]]:
         ("api", "require_confirmation_dag_change"): ("webserver", "require_confirmation_dag_change", "3.1.0"),
         ("api", "instance_name"): ("webserver", "instance_name", "3.1.0"),
         ("dag_processor", "parsing_pre_import_modules"): ("scheduler", "parsing_pre_import_modules", "3.1.0"),
+        ("api", "log_config"): ("api", "access_logfile", "3.1.0"),
```
This is not safe to do - they are different config options. As mentioned in the comment above, this will use the old `access_logfile` value for `log_config` if it's set, which isn't what we want.

jedcunningham commented on airflow-core/newsfragments/52860.significant.rst:
```diff
@@ -0,0 +1,17 @@
+Replace API server ``access_logfile`` configuration with ``log_config``
+
+The API server configuration option ``[api] access_logfile`` has been replaced with ``[api] log_config`` to align with uvicorn's logging configuration instead of the legacy gunicorn approach.
+The new ``log_config`` option accepts a path to a logging configuration file compatible with ``logging.config.fileConfig``, providing more flexible logging configuration for the API server.
```
Is there a way we can continue to support a simple access logfile config? I just worry a bit about backcompat, right now this is a breaking change. Maybe we leave `access_logfile`, build our own `log_config` file dynamically, and make `log_config` and `access_logfile` mutually exclusive? And we should emit a deprecation warning when `access_logfile` is used.

jedcunningham replied:
Well, good and bad news. Good news is no backcompat issues to worry about! I believe it got broken in #45103. Bad news is I don't think `access_log` ever worked in AF3.
